src/main/java/com/amazon/opendistroforelasticsearch/knn/index/codec/KNN80Codec/KNN80DocValuesConsumer.java

```diff
@@ -81,55 +91,78 @@ public void addKNNBinaryField(FieldInfo field, DocValuesProducer valuesProducer)
+
+            String method = fieldAttributes.get(KNNConstants.KNN_METHOD);
+            if (method == null) {
+                throw new NullPointerException("Method cannot be null");
+            }
 
             /**
-             * We always write with latest NMS library version OR Faiss
+             * We always write with latest engine version
              */
             if (!isKnnLibLatest()) {
                 KNNCounter.GRAPH_INDEX_ERRORS.increment();
-                throw new IllegalStateException("KNN library version mismatch. Correct version: " +
-                                                "HnswLib: " + NmsLibVersion.VNMSLIB_2011.indexLibraryVersion() +
-                                                "Faiss: "   + FAISSLibVersion.VFAISS_165.indexLibraryVersion());
+                throw new IllegalStateException("KNN library version mismatch. Correct version: " + NMSLIB_NAME + ": "
+                        + NmsLibVersion.VNMSLIB_2011.indexLibraryVersion() + FAISS_NAME + ": "
+                        + FaissLibVersion.VFAISS_165.indexLibraryVersion());
             }
 
             /**
              * Make Engine Name Into FileName
              */
             BinaryDocValues values = valuesProducer.getBinary(field);
-            String hnswFileName = String.format("%s_%s_%s%s", state.segmentInfo.name, knnEngine.getLatestBuildVersion(),
-                    field.name, KNNCodecUtil.HNSW_EXTENSION);
+            String engineFileName = buildEngineFileName(state.segmentInfo.name, knnEngine.getLatestBuildVersion(),
+                    field.name, knnEngine.getExtension());
             String indexPath = Paths.get(((FSDirectory) (FilterDirectory.unwrap(state.directory))).getDirectory().toString(),
-                    hnswFileName).toString();
+                    engineFileName).toString();
 
             KNNCodecUtil.Pair pair = KNNCodecUtil.getFloats(values);
-            if (pair == null || pair.vectors.length == 0 || pair.docs.length == 0) {
-                logger.info("Skipping hnsw index creation as there are no vectors or docs in the documents");
+            if (pair.vectors.length == 0 || pair.docs.length == 0) {
+                logger.info("Skipping engine index creation as there are no vectors or docs in the documents");
                 return;
             }
 
             // Pass the path for the nms library to save the file
             String tempIndexPath = indexPath + TEMP_SUFFIX;
             AccessController.doPrivileged(
-                    new PrivilegedAction<Void>() {
-                        public Void run() {
-                            if(faissindex) {
-                                KNNFaissIndex.saveIndex(pair.docs, pair.vectors, tempIndexPath, algoParams, spaceType);
-                            } else {
-                                KNNNmsLibIndex.saveIndex(pair.docs, pair.vectors, tempIndexPath, algoParams, spaceType);
+                    (PrivilegedAction<Void>) () -> {
```

Comment:
please avoid else/else if , it is difficult to follow.

Reply:
```
            AccessController.doPrivileged(
                    (PrivilegedAction<Void>) () -> {
                        if(KNNEngine.NMSLIB.getName().equals(knnEngine.getName())) {
                            KNNNmsLibIndex.save(pair.docs, pair.vectors, tempIndexPath, algoParams, spaceType,
                                    method);
                            return null;
                        }
                        if (KNNEngine.FAISS.getName().equals(knnEngine.getName())) {
                            String extraParametersString = fieldAttributes.getOrDefault(KNNConstants.EXTRA_PARAMETERS, null);
                            Map<String, Object> extraParameterMap = Collections.emptyMap();
                            if (extraParametersString != null) {
                                try {
                                    extraParameterMap = XContentFactory.xContent(XContentType.JSON).createParser(NamedXContentRegistry.EMPTY, DeprecationHandler.THROW_UNSUPPORTED_OPERATION, extraParametersString).map();
                                } catch (IOException e) {
                                    throw new IllegalStateException(e);
                                }

                            }

                            KNNFaissIndex.save(pair.docs, pair.vectors, tempIndexPath, extraParameterMap, spaceType,
                                    method, trainingDatasetSizeLimit, minimumDatapoints);
                            return null;
                        } 
                        throw new IllegalStateException("Invalid engine: "+knnEngine.getName());
                    }
            );

Reply:
Sure, will update.